src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemEventArgs.cs

```diff
@@ -19,7 +19,7 @@ public FileSystemEventArgs(WatcherChangeTypes changeType, string directory, stri
         {
             _changeType = changeType;
             _name = name;
-            _fullPath = Combine(directory, name);
+            _fullPath = Path.GetFullPath(Combine(directory, name));
```

Comment:
Using `Path.GetFullPath` here will add an extra allocation every time `FileSystemWatcher` reports an event. We should consider caching the full path in FSW and use that to create a `FileSystemEventArgs`.

Reply:
Hi @Jozkee. Hope you had a nice holiday!

FSW will notify every time the file changes so the same allocation will happen again but in FSW this time if I am not 
mistaken. Unless you are referring to getting the fully qualified path of the **directory** (<code>_directory</code>private field) used to initialize FSW in the constructor.

If that is the case, then some of the test cases for <code>FileSystemEventrArgs</code> will be rendered obsolete because the constructor's _directory_ parameter will always be a fully qualified path.

This means also that we are pushing this responsibility in FSW instead of the FSEventArgs but I guess that is Ok since from what I saw <code>FileSystemEventArgs</code> is not used **standalone** anywhere else.

Reply:
> Unless you are referring to getting the fully qualified path of the directory (_directoryprivate field) used to initialize FSW in the constructor.

Yes I was thinking on something more like this. 

As it is now, it is calling Path.GetFullPath for each event reported by the same instance of FileSystemWatcher. We could cache the result of GetFullPath in `_directory` and pass that to `FileSystemEventArgs` and `RenamedEventArgs` through new internal constructors (that can skip path normalization).

Although before we proceed, there is a caveat with caching the full path of a FileSystemWatcher; currently `_directory ` holds a relative path (relative to the CWD), if the CWD changes, I think that the watched folder will change as well. I'm not sure if that's an scenario that we should continue to support or not. If we decide to proceed with the caching, that scenario will be broken.

@adamsitnik @carlossanlop @ericstj @danmoseley thoughts?

Reply:
@Jozkee I was thinking to keep the <code>_directory</code> as is and have another private filed <code>_directoryFullPath</code> that is written together with _directory (there are about 2-3 places). Then use the new full path dir only when instantiating the event args classes. Without interfering with how FSW works right now.

Reply:
That would also reflect the behavior change that I described above. e.g: FSW passes `_directory` to CreateFile
https://github.com/dotnet/runtime/blob/312c66f1fc2f749f56612999cb1adab9ca7fde59/src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Win32.cs#L31-L33

If we add _directoryFullPath there is a chance that `Path.GetFullPath(_directory) != _directoryFullPath` and hence `FileSystemEventsArgs.FullPath` would be incorrect.

Based on this discussion I suggest to not cache the full path and make FileSystemEventsArgs call GetFullPath always. What do you guys think?

Reply:
Resolving this comment as per: https://github.com/dotnet/runtime/pull/63051#discussion_r783507896